Fix missing ServiceCallSite.Key causing an unkeyed cache entry to be overwritten by a keyed instance

IliaShuliatikov opened the pull request:
Fix missing `ServiceCallSite.Key` causing an unkeyed cache entry to be overwritten by a keyed instance when using `DynamicServiceProviderEngine` and cover all cases (Generic ([TryCreateOpenGeneric](https://github.com/dotnet/runtime/blob/6e2274c87c699f6749473b83969f12949e48ec7d/src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/CallSiteFactory.cs#L202)) and Enumerable ([TryCreateEnumerable](https://github.com/dotnet/runtime/blob/6e2274c87c699f6749473b83969f12949e48ec7d/src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/CallSiteFactory.cs#L203)) and the basic one ([TryCreateExact](https://github.com/dotnet/runtime/blob/6e2274c87c699f6749473b83969f12949e48ec7d/src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/CallSiteFactory.cs#L201C45-L201C59)) ) with unit tests.

Fixes #111795

IliaShuliatikov committed:
Fix missing ServiceCallSite.Key causing an unkeyed cache entry to be overwritten by a keyed instance

IliaShuliatikov committed:
Change delay, fix typo, add new test covering the basic case

IliaShuliatikov committed:
Use Dynamic service provider in tests

steveharter commented on src/libraries/Microsoft.Extensions.DependencyInjection/tests/DI.Tests/ServiceProviderServiceExtensionsTest.cs:
```diff
@@ -276,6 +277,82 @@ private static IServiceProvider CreateTestServiceProvider(int count)
             return serviceCollection.BuildServiceProvider();
         }
 
+        [Fact]
```
Please change these 3 tests from `[Fact]` to `[OuterLoop] // Includes a long Task.Delay`

Also, I assume these 3 tests can repro without having the other pre-existing tests running.

IliaShuliatikov replied:
> Please change these 3 tests from [Fact] to [OuterLoop] // Includes a long Task.Delay

Done, added `OuterLoop` attribute to the new tests.

> Also, I assume these 3 tests can repro without having the other pre-existing tests running.

Yes, that's correct.

IliaShuliatikov committed:
Make new tests OuterLoop

steveharter commented:
> I feel that ServiceProviderExtensionsTest might not be the best class to contain the new tests, because they are intended to test the service container itself rather than the extension methods.

Perhaps the class `CallSiteTests` in file `CallSiteTests.cs` would be a better match?

IliaShuliatikov committed:
Move new tests to CallSiteTests and make them more specific to ServiceCallSite

IliaShuliatikov commented:
@KalleOlaviNiemitalo, @steveharter, that's a good point, thanks.

I moved the new tests to the `CallSiteTests` class and made them more specific to `ServiceCallSite` and removed the unnecessary dependency on `DynamicServiceProviderEngine`, please take a look.

steveharter reviewed (approved):
Thanks for fixing this.

IliaShuliatikov commented:
Glad I could help and make .NET a bit better!
